business/checkers/k8shttproutes/no_k8sgateway_checker.go

```diff
@@ -30,26 +33,46 @@ func (s NoK8sGatewayChecker) ValidateHTTPRouteGateways(validations *[]*models.Is
 	if len(s.K8sHTTPRoute.Spec.ParentRefs) > 0 {
 		for index, parentRef := range s.K8sHTTPRoute.Spec.ParentRefs {
 			if string(parentRef.Name) != "" && string(*parentRef.Kind) == kubernetes.K8sActualGatewayType && string(*parentRef.Group) == kubernetes.K8sNetworkingGroupVersionV1.Group {
-				namespace := s.K8sHTTPRoute.Namespace
+				gwNs := s.K8sHTTPRoute.Namespace
 				if parentRef.Namespace != nil && string(*parentRef.Namespace) != "" {
-					namespace = string(*parentRef.Namespace)
+					gwNs = string(*parentRef.Namespace)
 				}
-				valid = s.checkGateway(string(parentRef.Name), namespace, validations, fmt.Sprintf("spec/parentRefs[%d]/name/%s", index, string(parentRef.Name))) && valid
+				valid = CheckGateway(string(parentRef.Name), gwNs, s.K8sHTTPRoute.Namespace, s.Cluster, s.GatewayNames, s.Namespaces, validations, fmt.Sprintf("spec/parentRefs[%d]/name/%s", index, string(parentRef.Name))) && valid
 			}
 		}
 	}
 	return valid
 }
 
-func (s NoK8sGatewayChecker) checkGateway(name, namespace string, validations *[]*models.IstioCheck, location string) bool {
-	hostname := kubernetes.ParseGatewayAsHost(name, namespace)
-	for gw := range s.GatewayNames {
-		gwHostname := kubernetes.ParseHost(gw, namespace)
+func CheckGateway(gwName, gwNs, routeNs, cluster string, gatewayNames map[string]k8s_networking_v1.Gateway, nss models.Namespaces, validations *[]*models.IstioCheck, location string) bool {
+	hostname := kubernetes.ParseGatewayAsHost(gwName, gwNs)
+	for gw := range gatewayNames {
+		gwHostname := kubernetes.ParseHost(gw, gwNs)
 		if found := kubernetes.FilterByHost(hostname.String(), hostname.Namespace, gw, gwHostname.Namespace); found {
-			return true
+			if gwHostname.Namespace == routeNs {
+				return true
+			} else if IsGatewaySharedWithNS(routeNs, cluster, gatewayNames[gw], nss) {
+				return true
+			}
 		}
 	}
 	validation := models.Build("k8sroutes.nok8sgateway", location)
 	*validations = append(*validations, &validation)
 	return false
 }
+
+// If K8sGateway's allowedRoutes selector matches the labels of given namespace
+func IsGatewaySharedWithNS(namespace string, cluster string, gw k8s_networking_v1.Gateway, nss models.Namespaces) bool {
+	ns := nss.GetNamespace(namespace, cluster)
+	if ns == nil {
+		return false
+	}
+	for _, l := range gw.Spec.Listeners {
+		if *l.AllowedRoutes.Namespaces.From == "Selector" &&
```

Comment:
@hhovsepy I have just discovered another scenario that conflicts with the KIA1401 validation. According to https://gateway-api.sigs.k8s.io/concepts/api-overview/#restricting-route-attachment, the `namespaces.from` field supports the value `All`, which allow Routes from all Namespaces to be attached.

I used this value in the shared gateway, and the validation error arises when it shouldn't. We need to take this into account for the validation.

Reply:
good catch